Add support for quantized operator conversion from PT to C2 via ONNX

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29465 [WIP] JIT Pass to unpack quantized weights
* **#29169 Add support for quantized operator conversion from PT to C2 via ONNX**

Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

supriyar committed:
[WIP] Add support for quantized operator conversion from PT to C2 via ONNX
Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[WIP] Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[WIP] Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

houseroad commented on torch/onnx/symbolic_registry.py:
```diff
@@ -16,6 +17,17 @@
     module = importlib.import_module('torch.onnx.symbolic_opset{}'.format(opset_version))
     _symbolic_versions[opset_version] = module
 
+def register_quantized_ops(domain, version):
```
I would suggest to move this function to symbolic_caffe2, since this will only be used there.

houseroad commented on torch/onnx/utils.py:
```diff
@@ -655,6 +659,16 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
                 return symbolic_fn(g, *inputs, **attrs)
 
+        elif ns == "quantized":
+            if operator_export_type == OperatorExportTypes.ONNX_ATEN_FALLBACK:
```
Can we generalize this logic? If it's ONNX_ATEN_FALLBACK, you can set the domain to 'caffe2'. Otherwise, keep it ''

houseroad commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -0,0 +1,33 @@
+import torch.onnx
+# This import monkey-patches graph manipulation methods on Graph, used for the
+# ONNX symbolics
+import torch.onnx.utils
```
lint?

houseroad reviewed (commented):
This is a great first step. I would strongly suggest to add an end to end test to https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py.

Otherwise, we don't know whether this export is correct or not.

Plus some inline comments.

supriyar committed:
Update on "[WIP] Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

supriyar committed:
Update on "[WIP] Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

houseroad commented on torch/onnx/utils.py:
```diff
@@ -583,7 +583,12 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
+            import torch.onnx.symbolic_caffe2 as sym_caffe2
+            sym_caffe2.register_quantized_ops('caffe2', opset_version)
+        else:
+            sym_registry.register_version('', opset_version)
```
I think we should execute sym_registry.register_version('', opset_version) unconditionally.

houseroad commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -2438,5 +2440,54 @@ def setup_rnn_tests():
+                bias = torch.ones(2).to(torch.float)
+                out = qlinear(qx)
+                return out.dequantize()
+        x = torch.ones(2, 2, 2, 2)
```
Can we do some real computation, instead only do one by one? this may help us cover some bugs.

supriyar replied:
Right now we aren't converting to caffe2. This is just checking to make sure the onnx conversion flow doesn't throw error. We can revisit this test once we add onnx->c2 path for these ops as well.

houseroad commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -0,0 +1,57 @@
+    sym_help._quantized_ops.add(output)
+    return output
+
+def relu(g, input, scale, zero_point):
```
We need to cover the regular relu as well, right?

houseroad commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -0,0 +1,57 @@
+        if isfunction(op[1]) and not sym_registry.is_registered_op(op[0], domain, version):
+            sym_registry.register_op(op[0], op[1], domain, version)
+
+def linear_prepack(g, input, weight):
```
actually, for some ops, like relu, we need check whether inputs are quantized or not, and dispatch in the symbolics here. We should override the old symbolics since we already include the regular paths.

houseroad commented on torch/onnx/symbolic_opset11.py:
```diff
@@ -44,9 +43,32 @@ def pixel_shuffle(g, self, upscale_factor):
         return _unimplemented("pixel_shuffle", "only support 4d input")
     return g.op("DepthToSpace", self, blocksize_i=upscale_factor, mode_s="CRD")
 
+def interpolate_quantized(g, input, output_size, dim, interpolate_mode, align_corners=None):
```
Do we have quantized upsample op in ONNX? I think in ONNX it is just a regular operator which supports uint8 as inputs. So we can just call interpolate symbolic in interpolate_quantized.

supriyar replied:
The issue with upsample is that `input_size = input.type().sizes()` translates to None for quantized input. So for now I've set input size to a tensor of size 1 in interpolate_quantized and separated it out. We will have to handle it in the onnx->c2 conversion flow.

houseroad commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2039,3 +2039,21 @@ def group_norm(g, input, num_groups, weight, bias, eps, cudnn_enabled):
+        "Y_scale_f": scale,
+        "zero_point_i": zero_point,
+    }
+    output = g.op("_caffe2::Int8Quantize", input, **kwargs)
```
These two functions are caffe2 specific, let's move it to symbolic_caffe2.py first. Later we can add the official support for ONNX.

supriyar replied:
These ops are in the `aten::` namespace (not `quantized::` namespace) so they are registered using the default domain (symbolic_opset9).
I can add a comment here to update to official onnx op when supported.

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

dtcrong commented:
Has this finished?   I tried export Quant MobileNet module (in pytorch quant tutorial) but failed with error: not quant per tensor op.

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]

supriyar committed:
Update on "Add support for quantized operator conversion from PT to C2 via ONNX"


Summary:
This PR adds preliminary support required to be able to run quantized pytorch models on a C2 backend.
For quantized ops we use a custom domain name 'caffe2' to register the ops if they are in the "quantized" namespace.

Test Plan:
```
class MyFunction(nn.Module):
    @staticmethod
        def forward(x, y):
          qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8)
          qlinear = nnq.Linear(2, 2)
          weight = torch.quantize_per_tensor(y, 1.0, 0, torch.qint8)
          bias = torch.ones(2).to(torch.float)
          out = qlinear(qx)
          qconv = nnq.Conv2d(2, 1, kernel_size=2, stride=1, padding=0)
          out = qconv(out)
          return out.dequantize()

x = torch.ones(2,2,2,2)
y = torch.ones(2,2)
torch.onnx.export(MyFunction(), (x, y), "output.onnx", verbose=True, opset_version=9, operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```

Produces the following onnx output
```
graph(%0 : Float(2, 2, 2, 2),
      %1 : Float(2, 2)):
  %2 : Double() = onnx::Constant[value={1}](), scope: MyFunction
  %3 : Long() = onnx::Constant[value={0}](), scope: MyFunction
  %4 : QUInt8(2, 2, 2, 2) = _caffe2::Int8Quantize(%0, %2, %3), scope: MyFunction # onnx_symbolic.py:10:0
  %5 : Byte(104) = onnx::Constant[value= 2  2 [ Variable[CPULongType]{2} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:207:0
  %6 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedLinear
  %7 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedLinear
  %8 : QUInt8(2, 2, 2, 2) = _caffe2::Int8FC(%4, %5, %6, %7), scope: MyFunction/QuantizedLinear # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/linear.py:133:0
  %9 : Byte(128) = onnx::Constant[value= 1  2  2  2 [ Variable[CPULongType]{4} ]](), scope: MyFunction # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:111:0
  %10 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %11 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %12 : Tensor = onnx::Unsqueeze[axes=[0]](%10)
  %13 : Tensor = onnx::Unsqueeze[axes=[0]](%11)
  %14 : Tensor = onnx::Concat[axis=0](%12, %13)
  %15 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %16 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %17 : Tensor = onnx::Unsqueeze[axes=[0]](%15)
  %18 : Tensor = onnx::Unsqueeze[axes=[0]](%16)
  %19 : Tensor = onnx::Concat[axis=0](%17, %18)
  %20 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %21 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %22 : Tensor = onnx::Unsqueeze[axes=[0]](%20)
  %23 : Tensor = onnx::Unsqueeze[axes=[0]](%21)
  %24 : Tensor = onnx::Concat[axis=0](%22, %23)
  %25 : Long() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %26 : Double() = onnx::Constant[value={1}](), scope: MyFunction/QuantizedConv2d
  %27 : Long() = onnx::Constant[value={0}](), scope: MyFunction/QuantizedConv2d
  %28 : QUInt8(2, 1, 1, 1) = _caffe2::Int8Conv(%8, %9, %14, %19, %24, %25, %26, %27), scope: MyFunction/QuantizedConv2d # /home/supriyar/miniconda3/envs/pytorch/lib/python2.7/site-packages/torch/nn/quantized/modules/conv.py:133:0
  %29 : Float(2, 1, 1, 1) = _caffe2::Int8Dequantize(%28), scope: MyFunction # onnx_symbolic.py:20:0
  return (%29)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18329753](https://our.internmc.facebook.com/intern/diff/D18329753)

[ghstack-poisoned]
